### PR TITLE
C027 Issue 72 - layer name parsing

### PR DESCRIPTION
**Pull request by ale-cristofori**
## Description
This PR fixes an issue raised that has got to be fixed in 2021.01.xx  - 2021.02.xx and MapStore2 downstream project C027 (where it was originally raised from).

Implements an additional method to parse `csw` XML responses to grab the relevant information to display on the `Catalog` component in MapStore. Changes to the structure and specification of the returned XML document (introduced with  (RNDT / INSPIRE)) caused misfunctioning in the component.

1. It was not possible to add the WMS service from the CSW respective service on the base map.
2. In the metadata template of the item, the links titles to download the source data were showing as `undefined` rather than showing the name of the resource.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) N/A

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [x] Bugfix
 - [ ] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [x] Other... Please describe: Handles an additional CSW data response specification

## Issue

**What is the current behavior?**
User adds a CSW service endpoint returning XML based responses with the RNDT / INSPIRE specification, an example of such URL is the following:
`http://sr-vm107-sitgs1.comune.intranet:8080/geonetwork/srv/ita/csw` (can be accessed only if connected the customer VPN).

User saves the service settings with `Show Preview` and `Show Metadata Template` in `Advanced Settings`, the user provides the customer defined metadata template for the metadata. 

User clicks on the search button, returned records won't show the add '+' button on the top right corner of the record card. User expands the card component to show the extended information for the dataset on the metadata template. Links texts to download the source data (shp, kmz, etc.) are showing with the text `undefined`.

#[C027-#72](https://github.com/geosolutions-it/MapStore2-C027/issues/72)

**What is the new behavior?**
User adds a CSW service endpoint returning XML based responses with the RNDT / INSPIRE specification, an example of such URL is the following:
`http://sr-vm107-sitgs1.comune.intranet:8080/geonetwork/srv/ita/csw` (can be accessed only if connected the customer VPN).

User saves the service settings with `Show Preview` and `Show Metadata Template` in `Advanced Settings`, the user provides the customer defined metadata template for the metadata. 

User clicks on the search button, returned records will show the add '+' button on the top right corner of the record card where WMS getMap request template is correctly returned from the CSW service. User expands the card component to show the extended information for the dataset on the metadata template. Links texts to download the source data (shp, kmz, etc.) are showing with the text `Title of the resource - download format`.

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

## Other useful information

recording_1

### CODE REVIEW

File: web/client/utils/CatalogUtils.js [hunk 1]
```diff
@@ -64,6 +64,63 @@ const getThumb = (dc) => {
...
+    }
+    return false;
+};
+
+const getMetaDataDownloadFormat = (protocol = '', value = '') => {
+    let format = '';
+    const formatsMap = [
+        {
+            protocol: /www-download/g,
+            value: /shp\/.*\.zip/g,
```
Comment (offtherailz) [inline web/client/utils/CatalogUtils.js:92]: I like the final effect but unfortunately parsing the URL to guess the format is not a good practice, in general. If we have a protocol like WMS, we can expect the parameters "layers" etc..., but with this, we can not. Also having WMS in the URL is not synonymous of a WMS service. 
So this map should have, when `name` attribute is not present, only to check the protocol: 

- protocol = "http://www.opengis.net/def/serviceType/ogc/wms" --> displayValue: WMS
- protocol = "http://www.opengis.net/def/serviceType/ogc/wfs"  --> displayName: WFS
- protocol = "https://registry.geodati.gov.it/metadata-codelist/ProtocolValue/www-download" --> displayName: "Download"
- protocol = "http://www.w3.org/TR/xlink/" --> displayName "Link"
- default "Link"

I liked to have "shapeFile", kmz too, but we cannot assume anything from the URL.

File: web/client/utils/CatalogUtils.js [hunk 2]
```diff
@@ -79,7 +136,17 @@ const converters = {
...
+                        return uri.protocol && (
+                            /** wms protocol params are explicitly defined as attributes (INSPIRE)*/
+                            uri.protocol.match(/^OGC:WMS-(.*)-http-get-map/g) ||
+                            uri.protocol.match(/^OGC:WMS/g) ||
+                            /** wms protocol params must be extracted from the element text (RNDT / INSPIRE) */
+                            uri.protocol.match(/serviceType\/ogc\/wms/g));
+                    }));
+                }
+                if (wms && wms.protocol.match(/serviceType\/ogc\/wms/g) && wms.description === "access point") {
+                    wms = extractWMSParamsFromURL(wms);
```
Comment (offtherailz) [inline web/client/utils/CatalogUtils.js:149]: We prefer to not not add too many if/then that modify objects in a function, or with the evolution of the code we loose pieces and we can not undestand while reading the final role. This is a quick change that you should test. 
```suggestion
                            /** wms protocol params must be extracted from the element text (RNDT / INSPIRE) */
                            uri.protocol.match(/serviceType\/ogc\/wms/g) && extractWMSParamsFromURL(wms));
                    }));
                }
```

File: web/client/utils/CatalogUtils.js [hunk 3]
```diff
@@ -64,6 +64,63 @@ const getThumb = (dc) => {
...
+            protocol: /TR\/xlink/g,
+            value: /kml\/.*\.kmz/g,
+            displayValue: 'KML file'
+        }
+    ];
+    format = formatsMap.filter(formatItem => (formatItem.protocol ? protocol.match(formatItem.protocol) : true) && value.match(formatItem.value))[0]?.displayValue || protocol;
+    return format;
+};
+
+const getURILinkNames = (metadata, locales, uriItem) => {
```
Comment (offtherailz) [inline web/client/utils/CatalogUtils.js:114]: It returns links, not only link names. So getURILinks is a better name. 
```suggestion
const getURILinks= (metadata, locales, uriItem) => {
```

File: web/client/utils/CatalogUtils.js [hunk 4]
```diff
@@ -168,7 +235,7 @@ const converters = {
                 }
                 // parsing URI
                 if (dc && dc.URI && castArray(dc.URI) && castArray(dc.URI).length) {
-                    metadata = {...metadata, uri: ["<ul>" + castArray(dc.URI).map(u => `<li><a target="_blank" href="${u.value}">${u.name}</a></li>`).join("") + "</ul>"]};
+                    metadata = {...metadata, uri: ["<ul>" + castArray(dc.URI).map(getURILinkNames.bind(this, metadata, locales)).join("") + "</ul>"]};
```
Comment (offtherailz) [inline web/client/utils/CatalogUtils.js:238]:
```suggestion
                    metadata = {...metadata, uri: ["<ul>" + castArray(dc.URI).map(getURILinks.bind(this, metadata, locales)).join("") + "</ul>"]};
```

File: web/client/utils/__tests__/CatalogUtils-test.js [hunk 5]
```diff
@@ -541,6 +541,68 @@ describe('Test the CatalogUtils', () => {
...
+    });
+
+    it('csw with DC references with implicit name in wms URI (RNDT / INSPIRE) - download link - value only', () => {
+        const records = CatalogUtils.getCatalogRecords('csw', {
+            records: [{
+                dc: {
+                    title: 'Test Layer',
+                    URI: [
+                        {
+                            value: 'http://gisdata.provider.host/geoserver/wms?tiled=true&version=1.1.1'
```
Comment (offtherailz) [inline web/client/utils/__tests__/CatalogUtils-test.js:594]: here you need the protocol

File: web/client/utils/__tests__/CatalogUtils-test.js [hunk 6]
```diff
@@ -541,6 +541,68 @@ describe('Test the CatalogUtils', () => {
...
+                            TYPE_NAME: 'DC_1_1.URI',
+                            protocol: 'http://www.opengis.net/def/serviceType/ogc/wms',
+                            description: 'access point',
+                            value: 'http://geoserver/wms?SERVICE=WMS&VERSION=1.3.0&REQUEST=GetMap&LAYERS=workspace:layer'
+                        }
+                    ]
+                }
+            }]
+        }, {});
+        expect(records.length).toBe(1);
```
Comment (offtherailz) [inline web/client/utils/__tests__/CatalogUtils-test.js:559]: Here you have to test that the record contains wms reference in `references`?

**Review by offtherailz — CHANGES_REQUESTED** [review]

File: web/client/utils/CatalogUtils.js (hunk 1 above)
Comment (ale-cristofori) [reply web/client/utils/CatalogUtils.js:92]: @offtherailz , could `protocol = "http://www.w3.org/TR/xlink/" --> displayName "Link"` be omitted as the display value will be `Link`, the default one?

```
    const formatsMap = [
        {
            protocol: 'https://registry.geodati.gov.it/metadata-codelist/ProtocolValue/www-download',
            displayValue: 'Download'
        },
        {
            protocol: 'http://www.opengis.net/def/serviceType/ogc/wms',
            displayValue: 'WMS'
        },
        {
            protocol: 'http://www.opengis.net/def/serviceType/ogc/wfs',
            displayValue: 'WFS'
        },
    ];
```

File: web/client/utils/__tests__/CatalogUtils-test.js (hunk 5 above)
Comment (ale-cristofori) [reply web/client/utils/__tests__/CatalogUtils-test.js:594]: I am happy to put it, but this is designed to specifically test the case where the `<dc:URI>` entry comes without any attribute. One example is the layer 'Sottopassaggi' in the C027 project datasets. While backporting I noticed the issue and provided a test for this case.

File: web/client/utils/CatalogUtils.js [hunk 7]
```diff
@@ -79,8 +128,17 @@ const converters = {
...
+                        return uri.protocol && (
+                            /** wms protocol params are explicitly defined as attributes (INSPIRE)*/
+                            uri.protocol.match(/^OGC:WMS-(.*)-http-get-map/g) ||
+                            uri.protocol.match(/^OGC:WMS/g) ||
+                            /** wms protocol params must be extracted from the element text (RNDT / INSPIRE) */
+                            uri.protocol.match(/serviceType\/ogc\/wms/g));
+                    }));
                 }
+                /** wms protocol params must be extracted from the element text (RNDT / INSPIRE) */
+                wms = (wms && wms.protocol.match(/serviceType\/ogc\/wms/g) && extractWMSParamsFromURL(wms));
```
Comment (offtherailz) [inline web/client/utils/CatalogUtils.js:141]:
```suggestion
                            (wms.protocol.match(/serviceType\/ogc\/wms/g) && extractWMSParamsFromURL(wms));
                    }));
                }
```

**Review by offtherailz — APPROVED** [review]

**Comment by offtherailz** [comment]
@ElenaGallo, could you please test this on  DEV ? Thank you

**Pull request merged**